### PR TITLE
NTO: Support for MachineConfigs.

### PR DESCRIPTION
**Pull request by jmencak**
Changes:
  - documented support for MachineConfigs/MachineConfigPools.
  - minor fixes (array -> list)

### CODE REVIEW

**Comment by jmencak** [comment]
@ahardin-rh  PTAL

File: modules/custom-tuning-specification.adoc [hunk 1]
```diff
@@ -43,43 +44,77 @@ profile:
...
-- label: <label_name>     # node or pod label name
-  value: <label_value>    # optional node or pod label value; if omitted, the presence of <label_name> is enough to match
-  type: <label_type>      # optional node or pod type ("node" or "pod"); if omitted, "node" is assumed
-  <match>                 # an optional <match> array
+- label: <label_name> <1>
+  value: <label_value> <2>
+  type: <label_type> <3>
+  <match> <4>
 ----
+<1> Node or pod label name.
```
Comment (ahardin-rh) [inline modules/custom-tuning-specification.adoc:86]: s/pod/Pod (a few instances)

File: modules/custom-tuning-specification.adoc [hunk 2]
```diff
@@ -43,43 +44,77 @@ profile:
...
-item of the `<match>` array matches, the entire `<match>` array evaluates to
-`true`. Therefore, the array acts as logical OR operator.
-
-.Example
-
+item of the `<match>` list matches, the entire `<match>` list evaluates to
+`true`. Therefore, the list acts as logical OR operator.
+
+If `machineConfigLabels` is defined, MachineConfigPool based matching is turned on
+for the given `recommend:` list item. This is in addition (logical OR) to the matching
```
Comment (ahardin-rh) [inline modules/custom-tuning-specification.adoc:99]: s/This is in addition to/Logical OR is in addition to 
(I think that's what is meant here with the parentheticals)

File: modules/custom-tuning-specification.adoc [hunk 3]
```diff
@@ -43,43 +44,77 @@ profile:
...
+needs and can only be applied by creating a MachineConfig. This involves
+finding all MachineConfigPools with machineConfigSelector matching
+`<mcLabels>` and setting the profile `<tuned_profile_name>` on all nodes that
+match the MachineConfigPools' `nodeSelectors`.
+
+[IMPORTANT]
+====
+When using MachineConfigPool based matching, it is advised to group nodes with the same
+hardware configuration into the same MachineConfigPool. Not following this practice
+may result in tuned operands calculating conflicting kernel parameters for two or more nodes
```
Comment (ahardin-rh) [inline modules/custom-tuning-specification.adoc:112]: s/may result/might result

File: modules/custom-tuning-specification.adoc [hunk 4]
```diff
@@ -121,3 +156,50 @@ catch-all to set `openshift-node` profile, if no other profile with higher
...
+    name: openshift-node-custom
+
+  recommend:
+  - machineConfigLabels:
+      machineconfiguration.openshift.io/role: "worker-custom"
+    priority: 20
+    profile: openshift-node-custom
+----
+
+To minimize Node reboots, it is recommended to label the target Nodes with
```
Comment (ahardin-rh) [inline modules/custom-tuning-specification.adoc:185]: s/Node/node
s/it is recommended to label the target/label the target

**Comment by ahardin-rh** [comment]
@jmencak Just a few minor comments from me. We'll also need to squash the changes down to 1 commit before we can merge. Thanks!

**Comment by ahardin-rh** [comment]
@mffiedler Can you please provide QE ack? Thanks!

**Comment by mffiedler** [comment]
@skordas Please review.   Thanks.

**Comment by skordas** [comment]
Great!
/lgtm

File: modules/custom-tuning-specification.adoc (hunk 2 above)
Comment (jmencak) [reply modules/custom-tuning-specification.adoc]: So this obviously needs to be explained better and I wonder if you can help here.  I had a go at this, comments welcome.

File: modules/custom-tuning-specification.adoc (hunk 1 above)
Comment (jmencak) [reply modules/custom-tuning-specification.adoc]: Good point, thanks!

File: modules/custom-tuning-specification.adoc (hunk 4 above)
Comment (jmencak) [reply modules/custom-tuning-specification.adoc]: Good catch!

**Review by jmencak — COMMENTED** [review]
Thanks for the review, Ashley, much appreciated.  Will squash once the section about machineConfigLabels is approved.

File: modules/custom-tuning-specification.adoc [hunk 5]
```diff
@@ -43,43 +44,81 @@ profile:
...
-
-.Example
-
+item of the `<match>` list matches, the entire `<match>` list evaluates to
+`true`. Therefore, the list acts as logical OR operator.
+
+If `machineConfigLabels` is defined, MachineConfigPool based matching is turned on
+for the given `recommend:` list item. `<mcLabels>` specifies the labels
+for a MachineConfig. The MachineConfig is created automatically to apply
+additional host settings (for example, kernel boot parameters) profile `<tuned_profile_name>`
```
Comment (ahardin-rh) [inline modules/custom-tuning-specification.adoc:101]: The MachineConfig is created automatically to apply host settings, such as kernel boot parameters, for the profile `<tuned_profile_name>`. You can only apply these settings by creating a MachineConfig.

File: modules/custom-tuning-specification.adoc (hunk 2 above)
Comment (ahardin-rh) [reply modules/custom-tuning-specification.adoc]: @jmencak I made a suggestion. Thanks!

File: modules/custom-tuning-specification.adoc (hunk 5 above)
Comment (jmencak) [reply modules/custom-tuning-specification.adoc]: Thanks for the suggestion.  I've taken the first sentence.  The seconds is slightly misleading.  Users will not create MachineConfigs directly, NTO creates those for the users.  What I was trying to convey is that NTO is creating MachineConfigs for host settings that cannot/should_not be performed in any other way. This is useful mostly for scenarios requiring host reboots (e.g. setting boot-time kernel parameters).  I believe that inclusion of the second sentence would not help, rather the opposite.  My original sentence was too long and cumbersome.  Let's keep this brief.  If you have other suggestions for improvements, happy to add them.  This paragraph is key for users to understand!

**Comment by mffiedler** [comment]
@skordas Please review most recent updates

File: modules/custom-tuning-specification.adoc [hunk 6]
```diff
@@ -43,43 +44,80 @@ profile:
...
+If `machineConfigLabels` is defined, MachineConfigPool based matching is turned on
+for the given `recommend:` list item. `<mcLabels>` specifies the labels
+for a MachineConfig. The MachineConfig is created automatically to apply host settings, such as
+kernel boot parameters, for the profile `<tuned_profile_name>`. This involves
+finding all MachineConfigPools with machineConfigSelector matching
+`<mcLabels>` and setting the profile `<tuned_profile_name>` on all nodes that
+match the MachineConfigPools' `nodeSelectors`.
+
+The list items `match` and `machineConfigLabels` are connected by the logical OR operator.
+The `match` item is evaluated first in a short-circuit manner, therefore, if it evaluates to
```
Comment (ahardin-rh) [inline modules/custom-tuning-specification.adoc:107]: manner. Therefore,

File: modules/custom-tuning-specification.adoc (hunk 5 above)
Comment (ahardin-rh) [reply modules/custom-tuning-specification.adoc]: @jmencak This reads much clearer. Thanks!

**Comment by jmencak** [comment]
@skordas Could you please review the latest changes so that they make sense from the user's/QE point of view?  We need to get this merged for 4.5.  Many thanks.

**Comment by skordas** [comment]
Sorry for late response.
/lgtm

**Comment by jmencak** [comment]
@ahardin-rh anything else that is needed to get this merged?  Thanks!

**Comment by ahardin-rh** [comment]
@jmencak Looks like we're all set! Merging :shipit:

**Pull request merged**